### ISSUE TITLE
A better test runner API

### ISSUE THREAD

**Issue by nicojs** [stryker-mutator/stryker-api#4]
I want to discus a better test runner API. Right now it looks like this (removed jsdoc for clarity):

``` javascript
 // Omitted some details for clarity

namespace TestRunnerFactory {
  class TestRunnerFactory extends Factory<RunnerOptions, TestRunner> {
  }
  let testRunnerFactoryInstance = new TestRunnerFactory();
  export function instance() {
    return <Factory<RunnerOptions, TestRunner>>testRunnerFactoryInstance;
  }
}

interface TestRunner {
  run(options: RunOptions): Promise<RunResult>;
}

interface RunResult {
  result: TestResult;
  testNames?: string[];
  succeeded?: number;
  failed?: number;
  timeSpent?: number;
  coverage?: CoverageCollection;
  errorMessages?: string[];
}

enum TestResult {
  Complete,
  Error,
  Timeout
}
```

I have a few problems with this. 
#### 1. We cannot see which tests failed and which tests passed.

When we have failed tests, we cannot see which of them failed and which of them passed. We also cannot see which error message belonged to which test. We could solve this by creating a new entity to hold the result of one single test. Something like this:

``` typescript
interface RunResult {
  testResults: TestResult[];
  timeSpent: number;
  coverage?: CoverageCollection;
}

interface TestResult{
  name: string;
  state: TestState;
  errorMessages?: string[];
}

enum TestState { Complete, Error, Timeout }
```

This can work for karma and mocha, which both know which tests are being ran. Which brings me to my next point.
#### 2. Implementing a test runner is kind of difficult.

When implementing a test runner you have to provide a lot of information. You need to report code coverage (it will work without it, but there is know way to communicate this to Stryker in advance), you need to report on the tests being ran and the state of each test with possible error. This is too bad, if you for example don't use an out-of-the-box test runner at all, but your custom testing logic.

What would the best course of action be here? We could change the interface of the `TestRunnerFactory` , in that it requires a static method `capabilities(): TestCapabilities` on each test runner. The test capabilities could look like this:

``` typescript
interface TestCapabilities {
  codeCoverage?: boolean;
  identitfyTests?: boolean;
}
```

Stryker could then use this interface to determine the best strategy to run mutations.

This brings me to my third point:
#### 3. Implementing code coverage reporting is a lot of work.

Implementing code coverage reporting is a lot of work. Just think about it. A test runner needs to know which code it must instrument for code coverage (the once with `mutated: true`), then it needs to instrument the code for code coverage and it has to do that with istanbul (or transform the code coverage report of the coverage tool in question). Finally, it needs to report the code coverage object. 

 Could we find a better way to implement code coverage? Maybe let Stryker instrument the code and then leave it up to the test runner to report the code coverage report (and reset it for next run). Maybe not ideal, as the code coverage responsibility gets filled by a tandem of Stryker and the test runner.

Can anyone think of a solution?

**Comment by simondel** [comment]
#### 1. We cannot see which tests failed and which tests passed.

I absolutely want this, but how will this work if a test runner does not know which test failed only that _X_ tests failed?

#### 2. Implementing a test runner is kind of difficult.

:+1: :+1: :+1: 

#### 3. Implementing code coverage reporting is a lot of work.

I don't think that every test runner will be able to the code would instrument for them (using Istanbul for example) without misusing the API of the underlying test runner (hacking it in there in a possibly unstable way). 

I think it would be a great option if this was a capability so a test runner could ask Stryker to instrument the code or signal that it will do it all itself

**Comment by nicojs** [comment]
> I absolutely want this, but how will this work if a test runner does not know which test failed only that X tests failed?

Well.. that's the problem. We can support both interfaces, but does that simplify it? I don't really know of a better  way though.

> What would the best course of action be here? We could change the interface of the TestRunnerFactory , in that it requires a static method `capabilities(): TestCapabilities` on each test runner.

About the `identitfyTests` capability. Is this really necessary? Or do we just run tests and roll with whatever the outcome of a test run is? Because a test runner might not know if it will be able to report individual test results. For example, a test runner might be able to report individual test names, but only when it runs testFramework X for example. So the only capability remaining is the `codeCoverage` capability.

> I don't think that every test runner will be able to the code would instrument for them (using Istanbul for example) without misusing the API of the underlying test runner (hacking it in there in a possibly unstable way). 

What do you mean with this?

I still want to do some more research on how Istanbul code coverage instrumentation works and how we would implement this in Stryker using both test runners we have know. I'll report here what i find.

**Comment by nicojs** [comment]
I did some more research on Istanbul and i think we can fairly simple to use the more low-level apis and let stryker be responsible for code coverage.

For example:

``` typescript
import {Instrumenter, Collector} from 'istanbul';
import * as fs from 'fs';

let instrumenter = new Instrumenter({coverageVariable: '__cov__'}); // default
let collector = new Collector();

let instrumented = instrumenter.instrumentSync('function mesaningOfLife() { return 42; }'
 /*This can even be the AST, this will save some time*/, 'filename.js');
fs.writeFileSync('filename.js', instrumented);
require('./filename'); // load the newly created file
collector.add(__cov__); // `__cov__` now magically exists
console.log(collector.getFinalCoverage())
```

As you can see, coverage consists of 2 steps. Instrument and Collect. Instrumentation can be done by any process, however the collection _only_ works when the tests are ran in the same node process. This means that we can only collect the report inside child process (`IsolatedTestRunnerAdapterRunner`) and we can than only use it for the mocha runner (and other node-based runners). For the karma runner and other browser based runners we could still instrument the code, but the runner should in that case be responsible for retrieving the `__cov__` global object.

In other words, i think the test capabilities interface should look like:

``` typescript
interface TestCapabilities {
  codeCoverageResponsibility?: CodeCoverageResponsibilityKind; // if empty, code coverage is disabled completely
}

enum CodeCoverageResponsibilityKind {
   // Test runner has full responsibility, stryker does nothing. Like the way stryker-karma-runner works now
  runner,
 // Test runner does nothing, stryker takes full responsibility. Should work for the stryker-mocha-runner
  stryker,
// Stryker instruments the code and the test-runner collects it. This could in theory make the stryker-karma-runner a lot faster as it does not have to re-instrument the code every time. Maybe we'll implement this in the future, not right away.   
tandem 
}
```

@simondel Do you (1) understand and (2) agree with this reasoning?

**Comment by nicojs** [comment]
Maybe these are better names:

``` typescript
CodeCoverageResponsibilityKind: {
InstrumentAndCollect,  Collect, None
}
``
```

**Comment by simondel** [comment]
What I meant with hacking things into Istanbul was that there is no really clean way to use it with our use case. We have to read files manually, pass that code to Istanbul etc. instead of just giving Istanbul a list of file names and letting it instrument them.

I understand the reasoning behind your coverage collection system with 3 different modes, but what if the test runner cannot collect and/or instrument code and if Stryker's system cannot be used (eg: a Karma runner which doesn't support code coverage)

**Comment by nicojs** [comment]
> I understand the reasoning behind your coverage collection system with 3 different modes, but what if the test runner cannot collect and/or instrument code and if Stryker's system cannot be used (eg: a Karma runner which doesn't support code coverage)

Well, in that case i think code coverage cannot be used. Maybe `CodeCoverageResponsibilityKind` can be `Off` in that case? Or just `null`? I also want to give users of Stryker the option to turn code coverage analysis off. I can think of scenario's where you would want to turn it off, for example in a small test suite. @simondel What do you think?

**Comment by simondel** [comment]
I think it's a good idea to allow people to turn off code coverage as it could indeed slow Stryker down in some cases (I think).

**Comment by nicojs** [comment]
Ok, we will have to make even more changes.

See issue [stryker issue#147](https://github.com/stryker-mutator/stryker/issues/147). The way of measuring code coverage right now involves having an entire test run per test, while only selecting one test at a time. We already knew this was kind of slow, but beside that, it also is memory consuming to the point that you get an out-of-memory error. This is unacceptable.

Good news: i think i found a way to achieve coverage-per-test results with only a single pass. See the latest commit of [karma-coverage-per-test](http://github.com/nicojs/karma-coverage-per-test). The way it works is to use a custom jasmine reporter. 

We first instrument the files for code coverage. We use a global variable other than `__coverage__` because that one is reserved for the code coverage _per test_ results. Lets say `__coverageCurrentSpecStryker__`. After each spec, we copy over the `__coverageCurrentSpecStryker__` results to the global `__coverage__[id]`, where `id` is the test identifier.

This is already enough for node-based test runners, but for karma, we still need to make sure karma will report this to the host stryker process. This works with a custom karma plugin. In a way, its even better, because it removes our dependency on `karma-coverage`. We inject our own plugin which reports back to stryker. One side note: this only works if we indeed use the global `__coverage__` object to store the results per test. This is a _magic_ variable which is pushed by all karma testFramework plugins (see [jasmine](https://github.com/karma-runner/karma-jasmine/blob/master/src/adapter.js#L190) and [mocha](https://github.com/karma-runner/karma-mocha/blob/master/src/adapter.js#L104) as examples)

So for this to work we need the following changes to the proposal:
- Stryker should do **all** the instrumentation. No preprocessing allowed by karma, because that polutes the global `__coverage__` object, which we absolute need.
- The karma test runner should communicate to us that it wants `CodeCoverageResponsibilityKind.Collect`. We remove `CodeCoverageResponsibilityKind.InstrumentAndCollect` as that is no longer allowed.
- We need to change the `TestSelector` interface to allow for custom code to run after each test. We should rename it to `TestFramework` as it is the better name now. For example:

``` typescript
interface TestFramework {
    select(id: string[]): string;
    afterEach(code: string): string;
}
```

The afterEach will return a fragment.

**Comment by nicojs** [comment]
Just to summarize: this would be the entire list of changes to the api.

Some new remarks:
1. We don't need `codeCoverageResponsibilityKind` after all. If files are instrumented, than the test runner should report it, or not if it runs tests in the same process (like the mocha-runner) because in that case Stryker can pick it up.
2. We rename ~~`TestSelector`~~ to `TestFramework` and thus remove the config for `TestSelector` so we need a new config property to disable code coverage analysis entirely.
3. `TestFramework`'s `select` is renamed to `filter`, makes more sense.
4. `TestFramework` should also have a `beforeEach` (next to `filter` and `afterEach`).
5. `CoverageResult` should now be reported per test, as we are collecting it in 1 pass of the test runner. We only need the `s` property per file, rest can be removed (performance).
6. `TestResult` should now also have a `timeSpentMs` as we do the initial run in one pass.

``` typescript

interface TestRunner extends EventEmitter { // <-- Add EventEmitter so it can emit 'testDone'
  run(options: RunOptions): Promise<RunResult>;
}

interface RunOptions {
  //   coverageEnabled?: boolean <-- Remove, not needed anymore as Stryker is doing the code coverage instrumentation
}

interface StrykerConfig {
   skipCoverageAnalysis: boolean; // <-- Add
   // testSelector: string; <-- remove
}

interface TestFramework { // renamed TestSelector
    filter(id: string[]): string;
    beforeEach(code: string): string;
    afterEach(code: string): string;
}

interface RunResult {
  tests: TestResult[];
  errorMessages?: string[];
  state: RunState;
}

interface TestResult {
  name: string;
  state: TestState;
  errorMessages?: string[];
  timeSpentMs?: number;
  coverage?: CoverageResult;
}

interface CoverageResult {
   [file: string]: {
       s: CoverageData;
       // Others we don't need
   }
}

enum RunState { Complete, Error, Timeout }
enum TestState { Success, Failed, Skipped }
```

**Comment by nicojs** [comment]
Just one more addition: i think we should remove this one: `skipCoverageAnalysis: boolean`. Instead:

``` typescript

  /**
   * Indicates which coverage analysis strategy to use.
   * During mutation testion, stryker will try to only run the tests that cover a particular line of code.
   * 
   * 'perTest' (default): Analyse coverage per test.
   * 'all': Analyse the coverage for the entire test suite.
   * 'off': Don't use coverage analysis 
   */
  coverageAnalysis?: 'perTest' | 'all' | 'off';
```

**Comment by nicojs** [comment]
Implemented this api in branch [feat-better-testrunner-api](https://github.com/stryker-mutator/stryker-api/tree/feat-better-testrunner-api)

**Comment by nicojs** [comment]
Another addition: we moved the coverage report up to the test runner in the latest commit.

``` typescript
interface RunResult {
  coverage?: CoverageCollection | CoverageCollectionPerTest;
}

interface CoverageCollectionPerTest {
  [testId: number]: CoverageCollection;
}
```

This way, it is easier to implement test runners. They only need to put the global `__coverage__` object in the `coverage` property of the `RunResult`

**Comment by nicojs** [comment]
Fixed with #6
